crates/autopilot/src/run_loop.rs

```diff
@@ -405,20 +409,62 @@ impl RunLoop {
 
     /// Execute the solver's solution. Returns Ok when the corresponding
     /// transaction has been mined.
-    async fn settle(&self, driver: &infra::Driver, solved: &Solution) -> Result<(), SettleError> {
+    async fn settle(
+        &self,
+        driver: infra::Driver,
+        solved: &Solution,
+        auction_id: i64,
+    ) -> Result<(), SettleError> {
         let order_ids = solved.order_ids().copied().collect();
         self.persistence
             .store_order_events(order_ids, OrderEventLabel::Executing);
 
+        let eth = self.eth.clone();
+        let persistence = self.persistence.clone();
+        let max_blocks_wait = self.max_blocks_wait;
+        let (cancellation_tx, cancellation_rx) = watch::channel(false);
+        let settlement_tx_wait_task = tokio::spawn(async move {
+            let tag = auction_id.to_be_bytes();
+            Self::wait_for_settlement_transaction(
+                eth,
+                persistence,
+                &tag,
+                max_blocks_wait,
+                cancellation_rx,
+            )
+            .await
+            .map_err(SettleError::Failure)
+        });
+
+        let duration = self.max_settlement_transaction_wait;
         let request = settle::Request {
             solution_id: solved.id,
         };
-
-        let tx_hash = driver
-            .settle(&request, self.max_settlement_transaction_wait)
-            .await
-            .map_err(SettleError::Failure)?
-            .tx_hash;
+        let driver_settle_task = tokio::spawn(async move {
+            driver
+                .clone()
+                .settle(request, duration)
+                .await
+                .map_err(SettleError::Failure)
+        });
+        // Wait for either the settlement transaction to be mined or the driver returned
+        // a result.
+        let tx_hash = tokio::select! {
+            transaction_result = settlement_tx_wait_task => {
+                transaction_result
+                    .map_err(|err| SettleError::Failure(err.into()))
+                    .and_then(|res| res.and_then(|tx| {
+                        tx.map(|tx| tx.hash).ok_or(SettleError::Failure(anyhow::anyhow!("settlement transaction await task reached deadline")))
+                    }))
+            },
+            settle_result = driver_settle_task => {
+                // When driver returns any result, the blocks fetching task has to be stopped.
+                let _ = cancellation_tx.send(true);
```

Comment:
Maybe we could add a warning here if the `send()` failed. As it is it is completely ignored :thinking:

Reply:
Even though we don't even care about the result, that might make sense to identify whether the cancellation logic has flaws.